ddtrace/contrib/internal/langchain/utils.py

```diff
@@ -0,0 +1,125 @@
+import inspect
+import sys
+
+
+class BaseTracedLangChainStreamResponse:
+    def __init__(self, generator, integration, span, on_span_finish):
+        self._generator = generator
+        self._dd_integration = integration
+        self._dd_span = span
+        self._on_span_finish = on_span_finish
+        self._chunks = []
+
+
+class TracedLangchainStreamResponse(BaseTracedLangChainStreamResponse):
+    def __enter__(self):
+        self._generator.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._generator.__exit__(exc_type, exc_val, exc_tb)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            chunk = self._generator.__next__()
+            self._chunks.append(chunk)
+            return chunk
+        except StopIteration:
+            self._on_span_finish(self._dd_span, self._chunks)
+            self._dd_span.finish()
+            raise
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            self._dd_integration.metric(self._dd_span, "incr", "request.error", 1)
+            raise
+
+
+class TracedLangchainAsyncStreamResponse(BaseTracedLangChainStreamResponse):
+    async def __aenter__(self):
+        await self._generator.__enter__()
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self._generator.__exit__(exc_type, exc_val, exc_tb)
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        try:
+            chunk = await self._generator.__anext__()
+            self._chunks.append(chunk)
+            return chunk
+        except StopAsyncIteration:
+            self._on_span_finish(self._dd_span, self._chunks)
+            self._dd_span.finish()
+            raise
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            self._dd_integration.metric(self._dd_span, "incr", "request.error", 1)
+            raise
+
+
+def shared_stream(
+    integration,
+    pin,
+    func,
+    instance,
+    args,
+    kwargs,
+    interface_type,
+    on_span_started,
+    on_span_finished,
+    **extra_options,
+):
+    options = {
+        "pin": pin,
+        "operation_id": f"{instance.__module__}.{instance.__class__.__name__}",
+        "interface_type": interface_type,
+    }
+
+    options.update(extra_options)
+
+    span = integration.trace(**options)
```

Comment:
Can we add a follow up item to submit llmobs spans from streamed responses?

Reply:
yep i have a ticket on our board!